include/dlaf/communication/rdma.h

```diff
@@ -44,7 +44,7 @@ namespace internal {
 /// Duplicates the tile to CPU memory if CUDA RDMA is not enabled for MPI.
 /// Returns the tile unmodified otherwise.
 template <Device D, typename T>
-auto prepareSendTile(hpx::shared_future<matrix::Tile<const T, D>> tile) {
+auto prepareSendTile(hpx::shared_future<matrix::Tile<T, D>> tile) {
```

Comment:
Not really. Following a discussion we had few weeks ago about this kind of stuff, for adapting the communication machinery for the GPU I opted for using `future::share()`, from which I get a `shared_future<Tile<T>>`.

This is potentially problematic because it opens to potential race-conditions, but as discussed, we limit this problem to this specific section of the code.

https://github.com/eth-cscs/DLA-Future/blob/b58654a6ed42734c5f32731a47817fbd1922c587/include/dlaf/communication/kernels/reduce.h#L82-L94

Maybe this code can be improved, so I will try to clarify a bit what I'm doing here (I will add/fix the documentation too).

The code mentioned covers different cases:
- INPUT: CPU vs GPU
- CPU-LAYOUT: contiguous vs non-contiguous

These are addressed in the following phases:
1. (compile-time) ensure that the input is available on CPU
2. (runtime) ensure that the CPU-LAYOUT is contiguous (temporary contiguous)
3. !! COMMUNICATION !!
4. (runtime) if temporary contiguous has been used, copy it back
5. (compile-time) if the input was on GPU, copy it back

The facility responsible for the 1st step is `prepareSendTile`, whose job is to return the same tile OR another temporary one, but whatever branch it takes, the returned tile will be on CPU.

If `prepareSendTile` would accept a `future`, it has to be moved in, so that:
- in the first case it can "pass-through" and get returned as it is;
- in the second case it will be just used as source of the copy.

But, in the latter case it means that we are going to loose the "handle" on the original tile, so we won't have it anymore for the 5th step where we will copy back results.

Sorry for the verbosity, but I wanted to explain (to myself) exactly what is going on in order to see if there may be room for improvements. Otherwise, I will just use this comment as reference for the documentation I'm going to write for the function.

Reply:
The explanation makes sense. I was more wondering about the general case, but given that this code might not survive too much longer we can take that when and if that change is needed. I just wanted to make sure I understood the change here.